net/net.go

```diff
@@ -997,10 +998,70 @@ func (n *net) putRecords(ctx context.Context, tid thread.ID, lid peer.ID, recs [
 		}
 	}
 
-	// setting new counters for heads
-	updatedCounter := head.Counter
-	connector, appConnected := n.getConnector(tid)
+	var (
+		connector, appConnected = n.getConnector(tid)
+		identity                = &thread.Libp2pPubKey{}
+		readKey                 *sym.Key
+		validate                bool
+		// setting new counters for heads
+		updatedCounter = head.Counter
+	)
+
+	if appConnected {
+		var err error
+		if readKey, err = n.store.ReadKey(tid); err != nil {
+			return err
+		} else if readKey != nil {
+			validate = true
+		}
+	}
+
 	for _, record := range chain {
+		if validate {
+			block, err := record.Value().GetBlock(ctx, n)
+			if err != nil {
+				return err
+			}
+
+			event, ok := block.(*cbor.Event)
+			if !ok {
+				event, err = cbor.EventFromNode(block)
+				if err != nil {
+					return fmt.Errorf("invalid event: %w", err)
+				}
+			}
+
+			dbody, err := event.GetBody(ctx, n, readKey)
+			if err != nil {
+				return err
+			}
+
+			if err = identity.UnmarshalBinary(record.Value().PubKey()); err != nil {
+				return err
+			}
+
+			if err = connector.ValidateNetRecordBody(ctx, dbody, identity); err != nil {
+				userErr := err
+
+				// remove stored internal blocks
+				header, err := event.GetHeader(ctx, n, nil)
+				if err != nil {
+					return err
+				}
+
+				body, err := event.GetBody(ctx, n, nil)
+				if err != nil {
+					return err
+				}
+
+				if err := n.RemoveMany(ctx, []cid.Cid{event.Cid(), header.Cid(), body.Cid()}); err != nil {
+					return fmt.Errorf("removing invalid blocks: %w", err)
+				}
+
+				return userErr
```

Comment:
See my comment at https://github.com/textileio/go-threads/pull/521/files#r642159397 maybe we can continue here instead of returning error? the caller can do nothing about the error apart from knowing the block being removed.

Reply:
Oop, sorry... I forgot about that comment. Copying it here:

> IIUC this change is to remove the invalid record but still return error, so the next call to it may succeed, but may still not if there are more records being invalid?

> In a general sense, this logic seems to be a bit unusual. Alternative ways could be either cleaning up all invalid records unconditionally and succeed, or returning a list of invalid records with errors so the caller can decide if it's okay to call some other method to "fix" them, but I could totally miss the context here.

If we continue instead of erroring, then we may end up with gaps, since a later record might pass the app-specific validation check. The logic here will cleanup the invalid record, and return an error to the caller that comes from the app-specific validation check, e.g., "write validation failed: only dogs are allowed, not cats". 

I could still be missing your point though!

Reply:
Ah okay so this `RemoveMany` is simply to clear the effect of adding this record which failed the app-specific validation check? Makes sense.

Reply:
Yep, that will remove all the blocks that are making up the single record.